src/modules/ecdsa_adaptor/main_impl.h

```diff
@@ -341,6 +341,13 @@ int secp256k1_ecdsa_adaptor_recover(const secp256k1_context* ctx, unsigned char
      * branch point. */
     secp256k1_declassify(ctx, &enckey_expected_ge, sizeof(enckey_expected_ge));
     if (!secp256k1_eckey_pubkey_serialize(&enckey_expected_ge, enckey_expected33, &size, SECP256K1_EC_COMPRESSED)) {
+        /* Unreachable
+         * Proof:
+         *     eckey_pubkey_serialize fails <=> deckey = 0
+         *     deckey = 0 <=> s^-1 = 0 or sp = 0
+         *     case 1: s^-1 = 0 impossible by the definition of multiplicative inverse
```

Comment:
While it's true that it's definitionally impossible, if `secp256k1_scalar_inverse` doesn't check for an overflow, then it can return `0` when `s` is equal to the curve order `q`. Neither the [implementation](https://github.com/ElementsProject/secp256k1-zkp/blob/f3708a1ecb445b1b05a0f8fcd1da6a88f83d89c4/src/scalar_impl.h#L68) on master nor the `scalar_low_impl.h` [implementation](https://github.com/ElementsProject/secp256k1-zkp/pull/132/files#diff-e04f40e44c1c6158458b5cfe25b01d18645b047d393d4bc5ae8c90f55d1803dfR130) implementation in this PR check for an overflow.

This PR adds a new `secp256k1_scalar_inverse` implementation in [scalar_4x64_impl.h](https://github.com/ElementsProject/secp256k1-zkp/pull/132/files#diff-334ffa39c358290caf0eadd32018e5d9276ea2e76416f28a34b981e59b56a9d9R933) and [scalar_8x32_impl.h](https://github.com/ElementsProject/secp256k1-zkp/pull/132/files#diff-4112c4b434fd39c8722830f9daf1db7b2feb89cfd73a69c57c4a575575ee9cedR811). This new implementation calls to either [secp256k1_scalar_to_signed62](https://github.com/ElementsProject/secp256k1-zkp/pull/132/files#diff-334ffa39c358290caf0eadd32018e5d9276ea2e76416f28a34b981e59b56a9d9R913) or [secp256k1_scalar_to_signed30](https://github.com/ElementsProject/secp256k1-zkp/pull/132/files#diff-4112c4b434fd39c8722830f9daf1db7b2feb89cfd73a69c57c4a575575ee9cedR786), which in turn check for an overflow (at https://github.com/ElementsProject/secp256k1-zkp/pull/132/files#diff-334ffa39c358290caf0eadd32018e5d9276ea2e76416f28a34b981e59b56a9d9R918 and https://github.com/ElementsProject/secp256k1-zkp/pull/132/files#diff-4112c4b434fd39c8722830f9daf1db7b2feb89cfd73a69c57c4a575575ee9cedR792 respectively), but only if the `VERIFY` macro is defined.

Reply:
But `s` doesn't overflow (it was just loaded from the ecdsa signature).

Reply:
The ecdsa signature can contain an `s` that is equal to the curve order `q`. When this happens in the implementation on master then `secp256k1_scalar_inverse` will write `0` to `deckey`, and the error at line 343 will be reached. With this PR, that is no longer the case because the `secp256k1_scalar_inverse` implementation has changed such that, if the `VERIFY` macro is defined then the following check will fail `VERIFY_CHECK(secp256k1_scalar_check_overflow(r) == 0);`.

I also just checked to see what happens if `VERIFY` is not defined, and with the new implementation `secp256k1_scalar_inverse` writes `1` to `deckey` instead of `0`, so even without `VERIFY`, the error at line 343 will not be reached in the new implementation. So it is correct to say that line 343 is unreachable with the new implementation (assuming `scalar_low_impl.h` is not used). But this depends on the implementation, and is not impossible, as is demonstrated by the test case on master which reaches line 343.

Reply:
Ok, I tweaked the comment such that it says that it's unreachable by the tests because scalar_inverse checks for valid inputs in that case (since VERIFY is defined).